backendRoot/src/main/java/com/virtuslab/gitmachete/backend/root/GitMacheteRepositoryBuilder.java

```diff
@@ -114,55 +116,84 @@ private GitMacheteRootBranch createGitMacheteRootBranch(IGitCoreRepository gitCo
     var pointedCommit = new GitMacheteCommit(corePointedCommit);
     var syncToOriginStatus = deriveSyncToOriginStatus(coreLocalBranch);
     var customAnnotation = entry.getCustomAnnotation().orElse(null);
-    var subbranches = deriveDownstreamBranches(gitCoreRepository, coreLocalBranch, entry);
+    AncestorityCache ancestorityCache = new AncestorityCache(gitCoreRepository);
+    var subbranches = deriveDownstreamBranches(gitCoreRepository, ancestorityCache, coreLocalBranch, entry);
 
     return new GitMacheteRootBranch(entry.getName(), subbranches, pointedCommit, syncToOriginStatus, customAnnotation);
   }
 
   private GitMacheteNonRootBranch createGitMacheteNonRootBranch(IGitCoreRepository gitCoreRepository,
+      AncestorityCache ancestorityCache,
       IGitCoreLocalBranch parentEntryCoreLocalBranch,
       BaseBranchLayoutEntry entry)
       throws GitMacheteException {
 
     IGitCoreLocalBranch coreLocalBranch = Try.of(() -> gitCoreRepository.getLocalBranch(entry.getName()))
         .getOrElseThrow(e -> new GitMacheteException(e));
 
-    Optional<BaseGitCoreCommit> coreForkPoint = Try.of(() -> coreLocalBranch.deriveForkPoint())
-        .getOrElseThrow(e -> new GitMacheteException(e));
+    Optional<BaseGitCoreCommit> deducedForkPoint = deduceForkPoint(ancestorityCache, coreLocalBranch,
+        parentEntryCoreLocalBranch);
 
     BaseGitCoreCommit corePointedCommit = Try.of(() -> coreLocalBranch.getPointedCommit())
         .getOrElseThrow(e -> new GitMacheteException(e));
 
-    if (!coreForkPoint.isPresent()) {
-      throw new GitMacheteException(String
-          .format("Cannot derive the fork point of non root branch \"%s\"", entry.getName()));
-    }
-
     // translate IGitCoreCommit list to IGitMacheteCommit list
-    List<IGitMacheteCommit> commits = Try.of(() -> coreLocalBranch.deriveCommitsUntil(coreForkPoint.get()))
-        .getOrElseThrow(e -> new GitMacheteException(e))
-        .map(GitMacheteCommit::new)
-        .collect(List.collector());
+    List<IGitMacheteCommit> commits = deducedForkPoint.isPresent()
+        ? Try.of(() -> coreLocalBranch.deriveCommitsUntil(deducedForkPoint.get()))
+            .getOrElseThrow(e -> new GitMacheteException(e))
+            .map(GitMacheteCommit::new)
+            .collect(List.collector())
+        : List.empty();
```

Comment:
Even when FP(A) is missing, when parent(A) is ancestor of A, we should fall back to the range of commits `parent(A)..A`

Reply:
> But what seems to be missing here is the fallback to parent(A) in case FP(A) is missing but A is descendant of parent(A)

But if the fallback to parent(A) takes place this piece of code works fine, am I right?

Reply:
Oh ok you're right, this should be fine now